src/main/java/org/prlprg/rds/RDSWriter.java

```diff
@@ -0,0 +1,552 @@
+    }
+  }
+
+  public void writeHeader() throws IOException {
+    // Could also be "B" (binary) and "A" (ASCII) but we only support XDR.
+    // XDR just means big endian and DataInputStream/DataOutputStream from Java use BigEndian
+    out.writeByte((byte) 'X');
+    out.writeByte((byte) '\n');
+
+    // Write version 2 of the encoding, since we want the writer to align with the reader and
+    // the reader does not support ALTREP
+    out.writeInt(2);
+
+    // Version of R for the writer
+    out.writeInt(RVersion.LATEST_AWARE.encode());
+
+    // Minimal version of R required to read back
+    out.writeInt((new RVersion(3, 5, 0, null)).encode());
+  }
+
+  public void write(SEXP sexp) throws IOException {
+    writeHeader();
+    writeItem(sexp);
+  }
+
+  // See
+  // https://github.com/wch/r-source/blob/65892cc124ac20a44950e6e432f9860b1d6e9bf4/src/main/serialize.c#L1021
+  public void writeItem(SEXP s) throws IOException {
+    var type = rdsType(s);
+
+    // Persisted through the ref table? TODO
+
+    switch (type) {
+      case RDSItemType.Special special -> {
+        // Save Special hooks: direct return and exit the function after a special hook
+        out.writeInt(special.i());
+      }
+      case RDSItemType.Sexp sexp -> {
+        // If refIndex is greater than 0, the object has already been written so we can
+        // write a reference (since the index is 1-based)
+        int sexpRefIndex = refTable.getOrDefault(s, 0);
+        if (sexpRefIndex > 0) {
+          writeRef(sexpRefIndex);
+          return;
+        }
+
+        // Otherwise, write the sexp as normal
+        switch (s) {
+          case RegSymSXP sym -> writeSymbol(sym);
+          case EnvSXP env -> writeEnv(env);
+
+            // Dotted-pair (pairlist) objects
+          case ListSXP list -> writeListSXP(list);
+          case LangSXP lang -> writeLangSXP(lang);
+          case PromSXP prom -> writePromSXP(prom);
+          case CloSXP clo -> writeCloSXP(clo);
+
+            // TODO: external pointer
+            // TODO: weak reference
+
+            // Other functions--special and builtin
+          case BuiltinOrSpecialSXP bos -> writeBuiltinOrSpecialSXP(bos);
+
+            // Vectors
+          case VectorSXP<?> vec -> writeVectorSXP(vec);
+
+            // Bytecode
+          case BCodeSXP bc -> writeByteCode(bc);
+
+          default -> throw new UnsupportedOperationException("Unsupported sexp type: " + s.type());
+        }
+      }
+    }
+  }
+
+  // utility functions ----------------------------------------------------------------------------
+
+  // Adds s to the ref table
+  private void refAdd(SEXP s) {
+    refTable.put(s, refIndex++);
+  }
+
+  // Determines if the hastag bit should be set based on the SEXP
+  private boolean hasTag(SEXP s) {
+    // ListSXP flags are handled elsewhere since they're written per-element
+    return switch (s) {
+      case CloSXP _clo -> true; // CloSXP should always be marked as having a tag I think
```

Comment:
Yes, that would be here in `serialize.c`: https://github.com/wch/r-source/blob/65892cc124ac20a44950e6e432f9860b1d6e9bf4/src/main/serialize.c#L1120 

For `PromSXP` (and `DotSXP`), we should check their tag attribute. The reader will check it (see `readPromise`) and then read the tag as the environment of the promise so in our case, checking if the environment of the promise is not `SEXP.EMPTY_ENV` should do it.
Tags for `LangSxp` are argument names (see https://cran.r-project.org/doc/manuals/r-devel/R-ints.html#SEXPTYPEs) but it does not seem we support them. I would suggest to let it as it is now but to add a comment here.